storage/integration_test.go

```diff
@@ -1609,6 +1571,132 @@ func TestIntegration_Objects(t *testing.T) {
 	})
 }
 
+func TestIntegration_Copy(t *testing.T) {
+	multiTransportTest(skipGRPC("allowlist issue potentially related to b/246634709"), t, func(t *testing.T, ctx context.Context, bucket string, prefix string, client *Client) {
+		bucketFrom := client.Bucket(bucket)
+		bucketInSameRegion := client.Bucket(prefix + uidSpace.New())
+		bucketInDifferentRegion := client.Bucket(prefix + uidSpace.New())
+
+		// Create new bucket
+		if err := bucketInSameRegion.Create(ctx, testutil.ProjID(), nil); err != nil {
+			t.Fatalf("bucket.Create: %v", err)
+		}
+		defer bucketInSameRegion.Delete(ctx)
+
+		// Create new bucket
+		if err := bucketInDifferentRegion.Create(ctx, testutil.ProjID(), &BucketAttrs{Location: "NORTHAMERICA-NORTHEAST2"}); err != nil {
+			t.Fatalf("bucket.Create: %v", err)
+		}
+		defer bucketInDifferentRegion.Delete(ctx)
+
+		// We use a larger object size to trigger multiple rewrite calls
+		minObjectSize := 2500000 // 2.5 Mb
+		obj := bucketFrom.Object("copy-object-original" + uidSpace.New())
+
+		// Create an object to copy from
+		w := obj.NewWriter(ctx)
+		c := randomContents()
+		for written := 0; written < minObjectSize; {
+			n, err := w.Write(c)
+			if err != nil {
+				t.Fatalf("w.Write: %v", err)
+			}
+			written += n
+		}
+		if err := w.Close(); err != nil {
+			t.Fatalf("w.Close: %v", err)
+		}
+
+		defer func() {
+			if err := obj.Delete(ctx); err != nil {
+				t.Errorf("obj.Delete: %v", err)
+			}
+		}()
+
+		attrs, err := obj.Attrs(ctx)
+		if err != nil {
+			t.Fatalf("obj.Attrs: %v", err)
+		}
+
+		crc32c := attrs.CRC32C
+
+		for _, test := range []struct {
+			desc        string
+			toObj       string
+			toBucket    *BucketHandle
+			copierAttrs *struct {
+				contentEncoding string
+			}
+		}{
+			{
+				desc:     "copy within bucket",
+				toObj:    "copy-within-bucket",
+				toBucket: bucketFrom,
+			},
+			{
+				desc:     "copy to new bucket",
+				toObj:    "copy-new-bucket",
+				toBucket: bucketInSameRegion,
+			},
+			{
+				desc:     "copy with attributes",
+				toObj:    "copy-with-attributes",
+				toBucket: bucketInSameRegion,
+				copierAttrs: &struct{ contentEncoding string }{
+					contentEncoding: "identity",
+				},
+			},
+			{
+				// this test should trigger multiple re-write calls and may fail
+				// with a rate limit error if those calls are stuck in an infinite loop
+				desc:     "copy to new region",
+				toObj:    "copy-new-region",
+				toBucket: bucketInDifferentRegion,
+			},
+		} {
+			t.Run(test.desc, func(t *testing.T) {
+				copyObj := test.toBucket.Object(test.toObj)
+				copier := copyObj.CopierFrom(obj)
+
+				copier.maxBytesRewrittenPerCall = 1048576
+
+				if attrs := test.copierAttrs; attrs != nil {
+					if attrs.contentEncoding != "" {
+						copier.ContentEncoding = attrs.contentEncoding
+					}
+				}
+
```

Comment:
Let's also specify Copier.ProgressFunc and check that the number of requests made is as expected (so we can ensure we are actually testing a multi-request copy).